tests/sample_size/scripts/test_utils.py

```diff
@@ -0,0 +1,274 @@
+            f"MDE: targeted treatment metric value minus the baseline value: "
+        )
+        mock_get_float.assert_called_once_with(test_mde, "minimum detectable effect")
+
+    @patch("sample_size.scripts.utils.input")
+    def test_get_variable_from_input_boolean(self, mock_input):
+        mock_input.return_value = " Boolean "
+
+        metric_type = utils.get_metric_type()
+
+        self.assertEqual(metric_type, "boolean")
+
+    @patch("sample_size.scripts.utils.input")
+    def test_get_variable_from_input_numeric(self, mock_input):
+        mock_input.return_value = " Numeric "
+
+        metric_type = utils.get_metric_type()
+
+        self.assertEqual(metric_type, "numeric")
+
+    @patch("sample_size.scripts.utils.input")
+    def test_get_variable_from_input_ratio(self, mock_input):
+        mock_input.return_value = " Ratio "
+
+        metric_type = utils.get_metric_type()
+
+        self.assertEqual(metric_type, "ratio")
+
+    @patch("sample_size.scripts.utils.input")
+    def test_get_variable_from_input_error(self, mock_input):
+        mock_input.return_value = "test"
+
+        with self.assertRaises(Exception) as context:
+            utils.get_metric_type()
+
+        self.assertEqual(
+            context.exception.args[0],
+            "Error: Unexpected metric type. Please enter Boolean, Numeric, or Ratio.",
+        )
+
+    @patch("sample_size.scripts.utils.get_float")
+    @patch("sample_size.scripts.utils.input")
+    def test_get_variable_parameters(self, mock_input, mock_get_float):
+        test_input_float = 5
+        mock_input.return_value = test_input_float
+        mock_get_float.return_value = test_input_float
+        test_parameter_definitions = {
+            "test": "test test",
+            "case": "case case",
+        }
+
+        result = utils.get_variable_parameters(test_parameter_definitions)
+
+        self.assertEqual(mock_input.call_count, 2)
+        self.assertEqual(mock_input.call_args_list[0][0][0], "Enter the test test: ")
+        self.assertEqual(mock_input.call_args_list[1][0][0], "Enter the case case: ")
+        self.assertEqual(mock_get_float.call_count, 2)
+        self.assertEqual(mock_get_float.call_args_list[0][0][0], test_input_float)
+        self.assertEqual(mock_get_float.call_args_list[0][0][1], "test test")
+        self.assertEqual(mock_get_float.call_args_list[1][0][0], test_input_float)
+        self.assertEqual(mock_get_float.call_args_list[1][0][1], "case case")
+        self.assertEqual(
+            result,
+            {
+                "test": test_input_float,
+                "case": test_input_float,
+            },
+        )
+
+    @patch("sample_size.scripts.utils.get_mde")
+    @patch("sample_size.scripts.utils.get_variable_parameters")
+    @patch("sample_size.scripts.utils.get_metric_type")
+    def test_get_metric_metadata_from_input(self, mock_get_metric_type, mock_get_variable_parameters, mock_get_mde):
```

Comment:
I wonder if we could approach testing this differently? In particular, if we could parameterize input as a number of test cases and focus exclusively on comparing the output of `get_metric_metadata_from_input` against some ground truth? Testing whether `mock_get_mde` was called, for example, might be an implementation detail that's better assessed by what MDE we return from `get_metric_metadata_from_input`.

Reply:
Can you be more specific on your suggestion? The functions like `get_mde` has their own separate tests. So they are tested for properly called here. Can you provide some examples?

Reply:
I've implemented a sketch of what I was describing in my own branch. Could you give me the permissions to be able to push my branch up to origin so I can link it here?

Reply:
@vjohnson1-godaddy you need to first be added to godaddy org to be the contributor, follow the instruction here https://github.secureserver.net/hivemind/docs/blob/master/onboarding/hivemind-developer-onboarding.md#opensource-contribution

Reply:
While I wait for my PR to be a contributor to be approved, I can clarify what I was talking about and provide some examples.

- Within the test `test_get_metric_metadata_from_input`, as I understand it, what we're really concerned about is that, given a set of inputs, we get a particular `metric_type` and `metric_metadata`.  The particular way that we implement that process could change, but as long as the same inputs produce the same outputs, the function as a whole is still valid. In the case of `mock_get_mde.assert_called_once_with(test_metric_type_lower)`, I can imagine valid scenarios where we change the implementation but the functionality remains the same. Perhaps `get_mde` could be replace, or have its function signature changed. In any case, we would want the test to pass regardless of changes to `get_med` _so long as the output is the same_.
- In this test, we only look at `Boolean` metrics, but we have some other metrics that we're concerned about as well. We could parameterize some of the input values like `mde` and check that all of the outputs match what we expect. That's what I meant by parameterization. In the example below, I left the role of defining what our expected output should be to snapshot testing, but one could also add the expected output as an additional parameter in the list of test cases passed to `@parameterized.expand(`

```python3

    @parameterized.expand(
        [
            # metric_type_input, parameters, mde,
            ("Boolean", {"probability": 0.01}, 0.05),
            ("Numeric", {"variance": 1}, 1),
            (
                "Ratio",
                {
                    "numerator_mean": 1,
                    "numerator_variance": 2,
                    "denominator_mean": 3,
                    "denominator_variance": 4,
                    "covariance": 5,
                },
                0.05,
            ),
        ]
    )
    @patch("sample_size.scripts.utils.get_mde")
    @patch("sample_size.scripts.utils.get_metric_parameters")
    @patch("sample_size.scripts.utils.get_metric_type")
    def test_get_metric_metadata(
        self, metric_type_input, parameters_input, mde_input, mock_get_metric_type, mock_get_metric_parameters, mock_get_mde
    ):
        mock_get_metric_type.return_value = metric_type_input
        mock_get_metric_parameters.return_value = parameters_input
        mock_get_mde.return_value = mde_input

        metric_type, metric_metadata = utils.get_metric_metadata()

        self.assertEqual(metric_type, metric_type_input.lower())
        self.snapshot.assert_match(  # type: ignore[attr-defined]
            json.dumps(metric_metadata, indent=2), f"{metric_type_input}.json"
        )
```

Reply:
seems like the script is trying to make sure `METRIC_PARAMETERS` is not changed. I'm not sure if it is necessary, since it is a constant. In `test_get_metric_parameters`, we already test that we will get correct parameters when provide correct parameter definitions like `METRIC_PARAMETERS`.

Reply:
Oh, I see, I see. Sounds reasonable!